Resolve animation issues with Date picker

disc0infern0 opened the pull request:
* Refactored date/time pickers to new view and updated control of state with new enum, allowing animations of picker view transitions to be handled precisely with new function setPickerState.
* No settings for transitions or animations work well (in a Form) when transitioning from a state where the date picker is shown, so use the setPickerState function to make transitions from a date picker shown state without animation, but animate all other view transitions.
* There is now no janky animation whatsoever, but 2 out of the 6 possible transitions now happen without animation.

disc0infern0 committed:
Convert viewModel instantiation from ObservedObject to StateObject in ReminderDetailsView

disc0infern0 committed:
Refactored Toggle and Date Time pickers from ReminderDetailsView into a new view; ReminderDateTimeView, and passed viewModel via the environment.

disc0infern0 committed:
ReminderDetailsViewModel: Bug fix for setPickerState.  Added withOptionalAnimation to honour Accessibility setting

disc0infern0 committed:
Refactored date/time pickers to new view and updated control of state with new enum, allowing animations of picker view transitions to be handled precisely with new function setPickerState.
No settings for transitions or animations work well (in a Form) when transitioning from a state where the date picker is shown, so use the setPickerState function to make transitions from a date picker shown state without animation, but animate all other view transitions.
There is now no janky animation whatsoever, but 2 out of the 6 possible transitions now happen without animation.

peterfriese commented on code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,8 @@
 
 /* Begin PBXBuildFile section */
+		49281E992760EBCB0046A465 /* ReminderDateTimeView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 49281E982760EBCB0046A465 /* ReminderDateTimeView.swift */; };
+		49281E9A2760EBCB0046A465 /* ReminderDateTimeView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 49281E982760EBCB0046A465 /* ReminderDateTimeView.swift */; };
```
Duplicate entries, might be a result of manually merging this file.

disc0infern0 replied:
agreed.

disc0infern0 replied:
On closer inspection, the two lines are not duplicates:  line 10 starts: 49281E99  vs line 11 starting: 49281E9A
The last character is different. Just to be sure, I ran 'uniq -d project.pbxproj' from the command line, and there was no output - indicating no duplicates.

Hminchae replied:
agreed

peterfriese commented on code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:
```diff
@@ -129,12 +134,11 @@ class ReminderDetailsViewModel: ObservableObject {
-      else {
-        hasDueTime = false
+        setPickerState(.date)
+      } else {
```
Personally, I am a fan of "egyptian style":

if condition {
  // do stuff
}
else { // <-- else is on new line, and (most importantly) parallel with the initial `if`
  // do other stuff
}

disc0infern0 replied:
Stylistically, I like that too, but generally I look to try to shorten my code as long as it remains clearly understood.
I am a fan of the old school Nassi Schneiderman style of coding, which I have adapted to mean that generally a function/view/discrete piece of code should all fit on the screen. If it doesn't, it needs refactoring.

disc0infern0 replied:
Appreciate your stylist conventions should take precedence in your repo!

peterfriese commented on code/frontend/MakeItSo/Shared/Features/ReminderList/ViewModels/ReminderDetailsViewModel.swift:
```diff
@@ -148,27 +152,45 @@ class ReminderDetailsViewModel: ObservableObject {
+  func setPickerState(_ newvalue: PickerState) {
+    // Dont animate if transitioning from state where date picker is shown
+    if pickerState == .date {
+      pickerState = newvalue
```
Use camel case: newValue

disc0infern0 replied:
agreed

peterfriese commented on code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:
```diff
@@ -0,0 +1,103 @@
+//
+//	ReminderDateTimeView.swift
```
Does this tab come from my template file...?

disc0infern0 replied:
Yes, I didn't change anything regarding templates, and that's what is generated from adding a new file.

peterfriese commented on code/frontend/MakeItSo/Shared/Features/ReminderList/Views/ReminderDateTimeView.swift:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  // Toggle label subviews for body
+  var dateToggleLabel: some View {
```
Can probably refactor this even more by turning into a function and passing in the icon and title.

Rename the function to something like `expandableRow` or `expandableRowLabel`, and replace `timeToggleLable` entirely.

disc0infern0 replied:
I thought about it, but you would to pass five parameters actually:
icon, title, dateValue, handler, textStyle

 HStack {
      Image(icon)
        .frame(width: 26, height: 26) //, alignment: .center)
        .background(.blue)
        .foregroundColor(.white)
        .cornerRadius(4)
      VStack(alignment: .leading) {
        Text(title)
        if dateValue != nil {
          Button(action: handler) {
            Text(dateValue, style: style)
        } ...

Not worth it for something that likely won't be re-used again?

peterfriese reviewed (changes requested):
A few comments regarding formatting, code structure, etc. Will take a closer look at the functionality later today.

disc0infern0 committed:
Updated code in ReminderDetailsViewModel.swift to satisfy Egyptian type if else formatting. Updated one instance of newvalue to newValue.

peterfriese reviewed (changes requested):
Just had an opportunity to check out how this looks like on the Simulator:

https://user-images.githubusercontent.com/232107/145356980-e555fd2b-ab81-4825-ae90-6b49874d7782.mp4

As becomes obvious from the screen recording, this is nowhere near the smooth transitions of the original app.

As long as we're not able to get close to the original, I am very reluctant to make wide-ranging changes to the code that will make it more difficult to read. Plus, there _is_ hope that Apple might actually fix this in the SDK itself, and I'd like to prevent us having to roll back a lot of changes we only made to workaround an issue.

Feel free to keep this PR open and continue optimising it, and use it as a reference when filing feedback to the SwiftUI team.

disc0infern0 committed:
ReminderListViewModel.
Removed unused Date functions and repositioned to end of file with additional commentary.
Moved definition of enum for PickerState into the class.

RemindersListViewModel.swift:
Removed unneeded delay in combine pipeline for removing blank entries.
